### PR TITLE
Evergreen controller

### PR DESCRIPTION
**Pull request by matthayter**
***Squash before merge***

Part of the Evergreen WIP: This PR introduces the controller, internal models and view models.

Remaining:
* Routes
* DB repository implementation of the new `VersionSpecificationRepository`

### CODE REVIEW

File: app/controllers/VersionSpecificationController.java [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+    }
+
+    /**
+     * Update or create a VersionSet with the given UUID.
+     *
+     * @param uuid The UUID.
+     * @return <code>Result</code> status code indicating success or failure.
+     */
+    public Result updateCreateVersionSet(final UUID uuid) {
+        final VersionSet view = request().body().as(VersionSet.class);
```
Comment (BrandonArp) [inline app/controllers/VersionSpecificationController.java]: Something tells me that this could go wrong.  Is view set to null? An exception thrown?

File: app/com/arpnetworking/metrics/portal/version_specifications/VersionSpecificationRepository.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+ * @author Matthew Hayter (mhayter at groupon dot com)
+ */
+public interface VersionSpecificationRepository {
+
+    /**
+     * Find the version set that corresponds to the set of host attributes by considering the list of version specifications.
+     *
+     * @param attributes The set of host attributes.
+     * @param ifModifiedSince An optional <code>DateTime</code> indicating that the result is not needed if the age
+     *                        of the underlying data is equal to or older than this moment. Leave this absent to disregard
```
Comment (BrandonArp) [inline app/com/arpnetworking/metrics/portal/version_specifications/VersionSpecificationRepository.java]: javadoc comments should not be indented/aligned like this

File: app/controllers/VersionSpecificationController.java (hunk 1 above)
Comment (vjkoskela) [inline app/controllers/VersionSpecificationController.java]: We generally avoid the use of the Play built-in JSON conversion routines and rely on our own ObjectMapper instance. Take a look at Deepika's expressions and alerts controllers.

File: app/controllers/VersionSpecificationController.java [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+    }
+
+    /**
+     * Retrieve the paginated list of <code>VersionSpecifications</code>.
+     *
+     * @param offset The number of version specifications to skip in the response
+     * @param limit The maximum number of version specifications to return.
+     * @return <code>Result</code> paginated list of version specifications.
+     */
+    public Result listVersionSpecifications(final Integer offset, final Integer limit) {
```
Comment (vjkoskela) [inline app/controllers/VersionSpecificationController.java]: The Integers in the signature should be declared @Nullable.

File: app/global/Rfc1123DateTimeParser.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+/**
+ * A parser and stringifier for the 'http datetime format' specified in RFC 822, updated by RFC 1123.
+ *
+ * @author Matthew Hayter (mhayter at groupon dot com)
+ */
+public class Rfc1123DateTimeParser {
+
+    private static final DateTimeFormatter RFC_1123_DATE_TIME_FORMATTER =
+            DateTimeFormat.forPattern("EEE, dd MMM yyyy HH:mm:ss 'GMT'").withZoneUTC().withLocale(Locale.US);
```
Comment (vjkoskela) [inline app/global/Rfc1123DateTimeParser.java]: Is there a reason you did not use the predefined formatter?

https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html

RFC_1123_DATE_TIME	RFC 1123 / RFC 822	'Tue, 3 Jun 2008 11:05:30 GMT'

File: app/global/Rfc1123DateTimeParser.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+import org.joda.time.format.DateTimeFormatter;
+
+import java.util.Locale;
+
+/**
+ * A parser and stringifier for the 'http datetime format' specified in RFC 822, updated by RFC 1123.
+ *
+ * @author Matthew Hayter (mhayter at groupon dot com)
+ */
+public class Rfc1123DateTimeParser {
```
Comment (vjkoskela) [inline app/global/Rfc1123DateTimeParser.java]: This seems to only wrap a date time formatter from java.time. Why not just use one directly?

File: app/models/internal/impl/DefaultVersionSet.java [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public int hashCode() {
+        return Objects.hashCode(_uuid, _version, _packageVersions);
+    }
+
+    @Override
```
Comment (vjkoskela) [inline app/models/internal/impl/DefaultVersionSet.java]: The default override javadoc should be applied to all overridden methods. I'm still open to you changing that convention, but it needs to be done wholesale and in a separate pr.

/**
 * {@inheritDoc}
 */

File: app/models/internal/impl/DefaultVersionSet.java [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+        @NotNull
+        @NotEmpty
+        private String _version;
+
+        @NotNull
+        private UUID _uuid;
+    }
+
+
+
```
Comment (vjkoskela) [inline app/models/internal/impl/DefaultVersionSet.java]: Unnecessary white space.

File: app/models/view/VersionSpecificationDetails.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String toString() {
+        return MoreObjects.toStringHelper(this)
+                .add("id", Integer.toHexString(System.identityHashCode(this)))
+                .add("class", this.getClass())
+                .add("uuid", _uuid)
```
Comment (vjkoskela) [inline app/models/view/VersionSpecificationDetails.java]: I believe we use capitals for all fields in toString; so this should be "Uuid".

File: test/global/Rfc1123DateTimeParserTest.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+ * limitations under the License.
+ */
+package global;
+
+import org.joda.time.DateTime;
+import org.junit.Assert;
+import org.junit.Test;
+
+/**
+ * @author Matthew Hayter (mhayter at groupon dot com)
```
Comment (vjkoskela) [inline test/global/Rfc1123DateTimeParserTest.java]: Missing class javadoc, although I suspect this will be deleted entirely.

File: test/global/Rfc1123DateTimeParserTest.java [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+ * @since 1.0.0
```
Comment (vjkoskela) [inline test/global/Rfc1123DateTimeParserTest.java]: I doubt that version is correct.

File: app/models/view/VersionSpecification.java [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String toString() {
+        return MoreObjects.toStringHelper(this)
+                .add("id", Integer.toHexString(System.identityHashCode(this)))
+                .add("class", this.getClass())
+                .add("uuid", _uuid)
```
Comment (vjkoskela) [inline app/models/view/VersionSpecification.java]: Uuid

File: app/models/internal/impl/DefaultVersionSpecificationAttribute.java [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public int hashCode() {
+        return Objects.hashCode(_key, _value);
+    }
+
+    @Override
```
Comment (vjkoskela) [inline app/models/internal/impl/DefaultVersionSpecificationAttribute.java]: Same note about the override documentation. Please make sure it is applied everywhere.

File: app/models/internal/impl/DefaultVersionSpecificationAttribute.java [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+    public String toString() {
+        return MoreObjects.toStringHelper(this)
+                .add("_key", _key)
```
Comment (vjkoskela) [inline app/models/internal/impl/DefaultVersionSpecificationAttribute.java]: Does not match conventional toString implementation.

File: app/models/internal/impl/DefaultVersionSpecification.java [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+     */
+    @Override
+    public int hashCode() {
+        return Objects.hashCode(_uuid, _versionSet, _versionSpecificationAttributes, _position);
+    }
+
+    @Override
+    public String toString() {
+        return MoreObjects.toStringHelper(this)
+                .add("_uuid", _uuid)
```
Comment (vjkoskela) [inline app/models/internal/impl/DefaultVersionSpecification.java]: Does not match conventional toString implementation.

File: app/models/internal/impl/DefaultVersionSet.java [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+    public int hashCode() {
+        return Objects.hashCode(_uuid, _version, _packageVersions);
+    }
+
+    @Override
+    public String toString() {
+        return MoreObjects.toStringHelper(this)
+                .add("_uuid", _uuid)
+                .add("_version", _version)
+                .add("_packageVersions", _packageVersions)
```
Comment (vjkoskela) [inline app/models/internal/impl/DefaultVersionSet.java]: Does not match conventional toString implementation.

File: app/models/internal/impl/DefaultPackageVersion.java [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+    @Override
+    public int hashCode() {
+        return Objects.hashCode(_name, _uri, _version);
+    }
+
+    @Override
+    public String toString() {
+        return MoreObjects.toStringHelper(this)
+                .add("_uri", _uri)
+                .add("_version", _version)
```
Comment (vjkoskela) [inline app/models/internal/impl/DefaultPackageVersion.java]: Does not match conventional toString implementation.

**Comment by vjkoskela** [comment]
Looks mostly like changes to adhere to style&convention. Please rebase and repost squashed. I don't think the free version of Github let's squash on merge.

File: app/global/Rfc1123DateTimeParser.java (hunk 4 above)
Comment (matthayter) [inline app/global/Rfc1123DateTimeParser.java]: :rage4: 

I think my problem was that I was looking for a joda formatter and skipped right over the Java8 native one. Does this project prefer one or the other? Seems to be joda...?

File: app/controllers/VersionSpecificationController.java (hunk 1 above)
Comment (matthayter) [inline app/controllers/VersionSpecificationController.java]: I can't see any example of parsing a JSON body in any other controller; indeed, every route in the routes file is `GET` based. Got any examples of using our own `ObjectMapper` in parsing a JSON body?

Comment (matthayter) [inline app/controllers/VersionSpecificationController.java]: Found one in `ElasticSearchHostRepository`:

    OBJECT_MAPPER.readValue(hit.getSourceAsString(), DefaultHost.class)
I guess I'll use that with `request().body().asText()`...

**Comment by matthayter** [comment]
Thanks @vjkoskela; all comments addressed

**Comment by matthayter** [comment]
Squashing, with no extra changes...
